docs/root/configuration/listeners/network_filters/smtp_proxy_filter.rst

```diff
@@ -0,0 +1,83 @@
+.. _config_network_filters_smtp_proxy:
+
+SMTP proxy
+==========
+
+The SMTP proxy filter decodes the initial capabilities negotiation
+between a SMTP client (downstream) and a SMTP server (upstream) to
+determine if the session will upgrade to TLS. It terminates TLS and
+then enters a "passthrough" mode like tcp_proxy.
+
+* SMTP :ref:`architecture overview <config_contrib_smtp>`
+
+.. attention::
+
+   The ``smtp_proxy`` filter is experimental and is currently under active development.
+   Capabilities will be expanded over time and the configuration structures are likely to change.
+
+
+Configuration
+-------------
+
+The SMTP proxy filter should be chained with TCP proxy and the
+StartTLS transport socket as shown in the configuration example below:
+
+.. code-block:: yaml
+
+    filter_chains:
+    - filters:
+      - name: envoy.filters.network.smtp_proxy
+        typed_config:
+          "@type": type.googleapis.com/envoy.extensions.filters.network.smtp_proxy.v3alpha.SmtpProxy
+          stat_prefix: smtp
+      - name: envoy.tcp_proxy
+        typed_config:
+          "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
+          stat_prefix: tcp
+          cluster: smtp_cluster
+
+      transport_socket:
+        name: envoy.transport_sockets.starttls
+        typed_config:
+          "@type": type.googleapis.com/envoy.extensions.transport_sockets.starttls.v3.StartTlsConfig
+          tls_socket_config:
+            # "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
+            common_tls_context:
+
+
+This can also be used with upstream_proxy_protocol on the upstream
+cluster to deliver the real client IP, etc, to an SMTP server that
+supports proxy protocol.
```

Comment:
How should the configuration look like when upstream cluster has upstream proxy protocol and upstream startssl?
Is it allowed?

Reply:
Let me test this locally, I don't remember whether I've tried this exact combination but I think it should work to chain upstream_proxy_protocol with starttls

Reply:
Filed https://github.com/envoyproxy/envoy/issues/29121 to track this as a known limitation though I think it's an unlikely configuration and referenced from the docs here.